### DISCUSSION TITLE
Corpus preparation and advice about training

### DISCUSSION THREAD

**Discussion by dorpxam** [lucasnewman/nanospeech#6] [Q&A]
First of all, thank you very much for this great project. Sincerely.

I want to train a French TTS voice based on a personal corpus recorded by myself after reading a bunch of great books. All the books was recorded using a Neumann Mic and anti-pop filter. The resulting quality is studio like : no noise, loudness consistency, high clarity. Here is the details of the corpus:

```
-------------------------------------------------------------------------------------------------------
| RAW FRENCH TTS CORPUS                                                                               |
-------------------------------------------------------------------------------------------------------
| ID |   DURATION   | TITLE                                                                           |
-------------------------------------------------------------------------------------------------------
| 01 |  18h 19m 41s | Dracula - Bram Stoker
| 02 |   8h 56m 24s | Fondation - Isaac Asimov
| 03 |   9h 40m 47s | Frankenstein ou le Prométhée Moderne - Mary Shelley
| 04 |  11h 34m 59s | La Nuit des Temps - René Barjavel
| 05 |   8h 47m 16s | Le Meilleur des Mondes - Aldous Huxley
| 06 |  10h 43m 11s | Le Portrait de Dorian Gray - Oscar Wilde
| 07 |   8h 55m 22s | Les Annales du Disque-Monde ：La Huitième Couleur (Tome 1) - Terry Pratchett
| 08 |   7h 48m 18s | Les Annales du Disque Monde ：Le Huitième Sortilège (Tome 2) - Terry Pratchett
| 09 |  16h 41m 54s | Les Hauts de Hurle-Vent - Emily Brontë
| 10 |  14h 48m 31s | Les Liaisons Dangereuses - Choderlos de Laclos
| 11 |  13h 02m 59s | Orgueil et Préjugés - Jane Austen
| 12 |   4h 17m 29s | Les Clous Rouges : Conan - Robert E. Howard
| 13 |   1h 15m 27s | Le Horla - Guy de Maupassant
| 14 |   1h 58m 34s | L'Iliade - Homère
| 15 |   2h 11m 38s | L'Odyssée - Homère
| 16 |   1h 59m 56s | Antigone - Jean Anouilh
| 17 |   1h 56m 02s | Andromaque - Jean Racine
| 18 |   0h 28m 48s | La vérité sur le cas de M. Valdemar - Edgar Allan Poe
| 19 |   0h 47m 11s | Ligeia - Edgar Allan Poe
| 20 |   1h 20m 48s | La Peur qui Rôde - H.P. Lovecraft
| 21 |   5h 55m 49s | Les Montagnes Hallucinées - H.P. Lovecraft
| 22 |   1h 36m 51s | Le Monstre sur le Seuil - H.P. Lovecraft
| 23 |   0h 27m 09s | La Révolte des Jouets - Philip K. Dick
| 24 |   0h 40m 53s | Le Canon - Philip K. Dick
| 25 |   0h 56m 57s | Le Crâne - Philip K. Dick
| 26 |   0h 14m 44s | Roug - Philip K. Dick
| 27 |   0h 39m 23s | Stabilité - Philip K. Dick
-------------------------------------------------------------------------------------------------------
| >> | 156h 07m 14s | 114 WAV Stereo 48000 Khz 16 bits | 100 GB                                       |    
-------------------------------------------------------------------------------------------------------
```
Before splitting the large WAV files into chuncks for training, I want to be sure that I will be in the right way. I've tuned `faster_whisper` to generate very good chuncks with high quality transcription using the model `bofenghuang/whisper-large-v3-french` (on HuggingFace).

This is the process I will use:
- Batch process the raw wav files using 'faster_whisper' into chuncks of 30 seconds
- For each chunck, record two files : [WAV-48Khz-16bits-Stereo] + [JSON] (text + precomputed duration)
- Batch process to resample the audio files to [WAV-24Khz-16bits-Mono] using Izotope RX SRC.
- Use your `create_sharded_wds_tar` process to build the TAR files.

Here's the few questions I have to clarify some points:
1) Is 156 hours enough for a single speaker voice ?
2) I do not want to use MP3 but instead keep the WAV format, it's OK ?
3) Your training code set the audio max duration to 10 seconds, can I use a max of 30 seconds of audio that clearly better match with sentences in my corpus ? In the experimentation section 5, you say "_Training on longer samples for better long-form generation_". I think 3 times longer is the case no ?

Last point. You say "_Extending the dataset and model architecture to guide generation with e.g. emotional expression or non-verbal utterances_". You mean, like [suno-ai/bark](https://github.com/suno-ai/bark) with the inclusion of commands like: `[laughs]` or `[clears throat]` ?

Thank you for your answer.

EDIT:
A little question about the numbers ? What is the best way to manage the number in the text JSON ? As literal or numeric ? both ? If both what is the best way to write both (something like "2|two") ?

**Comment by lucasnewman** [comment]
> Is 156 hours enough for a single speaker voice ?

It's difficult to say since "enough" is subjective for your use case — obviously bigger is better — but that's enough to get started and get the model speaking. Text alignment with speech and pronunciation will improve as you scale up both dataset size and compute.

> I do not want to use MP3 but instead keep the WAV format, it's OK ?

Yes, definitely. It gets turned into a Mel spectrogram in the pipeline, so as long as you can convert the source format to a waveform tensor, it's fine. Keep in mind you may need to tweak the dataloader slightly to find the right keys.

> Your training code set the audio max duration to 10 seconds, can I use a max of 30 seconds of audio that clearly better match with sentences in my corpus ? In the experimentation section 5, you say "Training on longer samples for better long-form generation". I think 3 times longer is the case no ?

Yep! I used 10 seconds since it's kind of a toy model and I wanted to train it quickly, but 30 seconds will work as long as you can fit it into memory.

> Last point. You say "Extending the dataset and model architecture to guide generation with e.g. emotional expression or non-verbal utterances". You mean, like [suno-ai/bark](https://github.com/suno-ai/bark) with the inclusion of commands like: [laughs] or [clears throat] ?

Yeah, that's the kind of thing I was referencing. There's also a really cool [paper](https://arxiv.org/pdf/2308.05725) and dataset by Meta that has emotional performances categorized with very high quality samples.

> A little question about the numbers ? What is the best way to manage the number in the text JSON ? As literal or numeric ? both ? If both what is the best way to write both (something like "2|two") ?

Since you're training it from scratch, you'd want to either 1) consistently normalize the text (like "2" -> "two") before you tokenize it or 2) just include it both ways and have it learn the equivalency as a data augmentation.
